File: py/test/selenium/webdriver/support/color_tests.py
# Licensed to the Software Freedom Conservancy (SFC) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The SFC licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from selenium.webdriver.support.color import Color


class TestColor(object):

    def test_rgb_to_rgb(self):
        rgb = "rgb(1, 2, 3)"
        assert Color.from_string(rgb).rgb == rgb

    def test_rgb_to_rgba(self):
        rgb = "rgb(1, 2, 3)"
        assert Color.from_string(rgb).rgba == "rgba(1, 2, 3, 1)"

    def test_rgb_pct_to_rgba(self):
        rgb = "rgb(10%, 20%, 30%)"
        assert Color.from_string(rgb).rgba == "rgba(25, 51, 76, 1)"

    def test_rgb_allows_whitespace(self):
        rgb = "rgb(\t1,   2    , 3)"
        assert Color.from_string(rgb).rgb == "rgb(1, 2, 3)"

    def test_rgba_to_rgba(self):
        rgba = "rgba(1, 2, 3, 0.5)"
        assert Color.from_string(rgba).rgba == rgba

    def test_rgba_pct_to_rgba(self):
        rgba = "rgba(10%, 20%, 30%, 0.5)"
        assert Color.from_string(rgba).rgba == "rgba(25, 51, 76, 0.5)"

    def test_hex_to_hex(self):
        hex_ = "#ff00a0"
        assert Color.from_string(hex_).hex == hex_

    def test_hex_to_rgb(self):
        hex_ = "#01Ff03"
        rgb = "rgb(1, 255, 3)"
        assert Color.from_string(hex_).rgb == rgb

    def test_hex_to_rgba(self):
        hex_ = "#01Ff03"
        rgba = "rgba(1, 255, 3, 1)"
        assert Color.from_string(hex_).rgba == rgba

        hex_ = "#00ff33"
        rgba = "rgba(0, 255, 51, 1)"
        assert Color.from_string(hex_).rgba == rgba

    def test_rgb_to_hex(self):
        assert Color.from_string("rgb(1, 255, 3)").hex == "#01ff03"

    def test_hex3_to_rgba(self):
        assert Color.from_string("#0f3").rgba == "rgba(0, 255, 51, 1)"

    def test_hsl_to_rgba(self):
        hsl = "hsl(120, 100%, 25%)"
        rgba = "rgba(0, 128, 0, 1)"
        assert Color.from_string(hsl).rgba == rgba

        hsl = "hsl(100, 0%, 50%)"
        rgba = "rgba(128, 128, 128, 1)"
        assert Color.from_string(hsl).rgba == rgba

    def test_hsla_to_rgba(self):
        hsla = "hsla(120, 100%, 25%, 1)"
        rgba = "rgba(0, 128, 0, 1)"
        assert Color.from_string(hsla).rgba == rgba

        hsla = "hsla(100, 0%, 50%, 0.5)"
        rgba = "rgba(128, 128, 128, 0.5)"
        assert Color.from_string(hsla).rgba == rgba

    def test_named_color(self):
        assert Color.from_string("green").rgba == "rgba(0, 128, 0, 1)"
        assert Color.from_string("gray").rgba == "rgba(128, 128, 128, 1)"
        assert Color.from_string("aqua").hex == "#00ffff"
        assert Color.from_string("transparent").rgba == "rgba(0, 0, 0, 0)"

    def test_equals(self):
        assert Color.from_string("#f00") == Color.from_string("rgb(255, 0, 0)")
        assert Color.from_string("rgba(30, 30, 30, 0.2)") != Color.from_string("rgba(30, 30, 30, 1)")

    def test_hash(self):
        hash1 = hash(Color.from_string("#f00"))
        hash2 = hash(Color.from_string("rgb(255, 0, 0)"))
        assert hash1 == hash2

    def test_string_representations(self):
        hex_ = "#01Ff03"
        assert str(Color.from_string(hex_)) == "Color: rgba(1, 255, 3, 1)"
        assert repr(Color.from_string(hex_)) == "Color(red=1, green=255, blue=3, alpha=1)"

File: dotnet/test/support/Colors/ColorTests.cs
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

namespace OpenQA.Selenium.Support.Colors
{
    [TestFixture]
    public class ColorTests
    {
        [Test]
        public void RgbToRgb()
        {
            string rgb = "rgb(1, 2, 3)";
            Assert.That(Color.FromString(rgb).Rgb, Is.EqualTo(rgb));
        }

        [Test]
        public void RgbToRgba()
        {
            string rgb = "rgb(1, 2, 3)";
            Assert.That(Color.FromString(rgb).Rgba, Is.EqualTo("rgba(1, 2, 3, 1)"));
        }

        [Test]
        public void RgbPctToRgba()
        {
            string rgba = "rgb(10%, 20%, 30%)";
            Assert.That(Color.FromString(rgba).Rgba, Is.EqualTo("rgba(25, 51, 76, 1)"));
        }

        [Test]
        public void RgbAllowsWhitespace()
        {
            string rgb = "rgb(\t1,   2    , 3)";
            string canonicalRgb = "rgb(1, 2, 3)";
            Assert.That(Color.FromString(rgb).Rgb, Is.EqualTo(canonicalRgb));
        }

        [Test]
        public void RgbaToRgba()
        {
            string rgba = "rgba(1, 2, 3, 0.5)";
            Assert.That(Color.FromString(rgba).Rgba, Is.EqualTo(rgba));
        }

        [Test]
        public void RgbaPctToRgba()
        {
            string rgba = "rgba(10%, 20%, 30%, 0.5)";
            Assert.That(Color.FromString(rgba).Rgba, Is.EqualTo("rgba(25, 51, 76, 0.5)"));
        }

        [Test]
        public void HexToHex()
        {
            string hex = "#ff00a0";
            Assert.That(Color.FromString(hex).Hex, Is.EqualTo(hex));
        }

        [Test]
        public void HexToRgba()
        {
            string hex = "#01Ff03";
            string rgba = "rgba(1, 255, 3, 1)";
            Assert.That(Color.FromString(hex).Rgba, Is.EqualTo(rgba));
            // same test data as hex3 below
            hex = "#00ff33";
            rgba = "rgba(0, 255, 51, 1)";
            Assert.That(Color.FromString(hex).Rgba, Is.EqualTo(rgba));
        }

        [Test]
        public void RgbToHex()
        {
            string hex = "#01ff03";
            string rgb = "rgb(1, 255, 3)";
            Assert.That(Color.FromString(rgb).Hex, Is.EqualTo(hex));
        }

        [Test]
        public void Hex3ToRgba()
        {
            string hex = "#0f3";
            string rgba = "rgba(0, 255, 51, 1)";
            Assert.That(Color.FromString(hex).Rgba, Is.EqualTo(rgba));
        }

        [Test]
        public void HslToRgba()
        {
            string hsl = "hsl(120, 100%, 25%)";
            string rgba = "rgba(0, 128, 0, 1)";
            Assert.That(Color.FromString(hsl).Rgba, Is.EqualTo(rgba));
            hsl = "hsl(100, 0%, 50%)";
            rgba = "rgba(128, 128, 128, 1)";
            Assert.That(Color.FromString(hsl).Rgba, Is.EqualTo(rgba));
            hsl = "hsl(0, 100%, 50%)"; // red
            rgba = "rgba(255, 0, 0, 1)";
            Assert.That(Color.FromString(hsl).Rgba, Is.EqualTo(rgba));
            hsl = "hsl(120, 100%, 50%)"; // green
            rgba = "rgba(0, 255, 0, 1)";
            Assert.That(Color.FromString(hsl).Rgba, Is.EqualTo(rgba));
            hsl = "hsl(240, 100%, 50%)"; // blue
            rgba = "rgba(0, 0, 255, 1)";
            Assert.That(Color.FromString(hsl).Rgba, Is.EqualTo(rgba));
            hsl = "hsl(0, 0%, 100%)"; // white
            rgba = "rgba(255, 255, 255, 1)";
            Assert.That(Color.FromString(hsl).Rgba, Is.EqualTo(rgba));
        }

        [Test]
        public void HslaToRgba()
        {
            string hsla = "hsla(120, 100%, 25%, 1)";
            string rgba = "rgba(0, 128, 0, 1)";
            Assert.That(Color.FromString(hsla).Rgba, Is.EqualTo(rgba));
            hsla = "hsla(100, 0%, 50%, 0.5)";
            rgba = "rgba(128, 128, 128, 0.5)";
            Assert.That(Color.FromString(hsla).Rgba, Is.EqualTo(rgba));
        }

        [Test]
        public void BaseColourToRgba()
        {
            string baseColour = "green";
            string rgba = "rgba(0, 128, 0, 1)";
            Assert.That(Color.FromString(baseColour).Rgba, Is.EqualTo(rgba));
            baseColour = "gray";
            rgba = "rgba(128, 128, 128, 1)";
            Assert.That(Color.FromString(baseColour).Rgba, Is.EqualTo(rgba));
        }

        [Test]
        public void TransparentToRgba()
        {
            string transparent = "transparent";
            string rgba = "rgba(0, 0, 0, 0)";
            Assert.That(Color.FromString(transparent).Rgba, Is.EqualTo(rgba));
        }

        [Test]
        public void CheckEqualsWorks()
        {
            Color objectA = Color.FromString("#f00");
            Color objectB = Color.FromString("rgb(255, 0, 0)");
            Assert.That(objectA, Is.EqualTo(objectB));
        }

        [Test]
        public void CheckHashCodeWorks()
        {
            Color objectA = Color.FromString("#f00");
            Color objectB = Color.FromString("rgb(255, 0, 0)");
            Assert.That(objectA.GetHashCode(), Is.EqualTo(objectB.GetHashCode()));
        }

        [Test]
        public void CheckSettingOpacityRGB()
        {
            string initial = "rgb(1, 255, 3)";
            Color actual = Color.FromString(initial);

            actual.Alpha = 0.5;

            string expected = "rgba(1, 255, 3, 0.5)";
            Assert.That(expected, Is.EqualTo(actual.Rgba));
        }

        [Test]
        public void CheckSettingOpacityRGBA()
        {
            string initial = "rgba(1, 255, 3, 1)";
            Color actual = Color.FromString(initial);

            actual.Alpha = 0;

            string expected = "rgba(1, 255, 3, 0)";
            Assert.That(expected, Is.EqualTo(actual.Rgba));
        }

        [Test]
        public void BaseColourToAwt()
        {
            System.Drawing.Color green = System.Drawing.Color.FromArgb(255, 0, 255, 0);
            string rgba = "rgba(0, 255, 0, 1)";
            Assert.That(Color.FromString(rgba).GetColor, Is.EqualTo(green));
        }

        [Test]
        public void TransparentColourToAwt()
        {
            System.Drawing.Color transGreen = System.Drawing.Color.FromArgb(0, 0, 255, 0);
            string rgba = "rgba(0, 255, 0, 0)";
            Assert.That(Color.FromString(rgba).GetColor, Is.EqualTo(transGreen));
        }
    }
}